sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2099,9 +2099,14 @@ class Analyzer(override val catalogManager: CatalogManager)
         ResolvedFunc(Identifier.of(funcIdent.database.toArray, funcIdent.funcName))
 
       case q: LogicalPlan =>
+        val isGroupingIdAllowed = q match {
+          case _: Aggregate | _: GroupingSets | _: Filter | _: Sort => true
```

Comment:
what if `grouping__id` refers to an actual column (e.g. from a table) in these operators?

Reply:
It will throw an analysis exception both after/before this PR, e.g.,
```
scala> sql("select * from t3").show()
+---+---+------------+
|  a|  b|grouping__id|
+---+---+------------+
|  1|  2|           3|
+---+---+------------+

// with/without this PR
scala> sql("select * from t3 where grouping__id > 0").show()
org.apache.spark.sql.AnalysisException: grouping()/grouping_id() can only be used with GroupingSets/Cube/Rollup;
```
On the other hand, this PR will change behaviours other than these operators above, e.g.,
```
// The current behaviour
scala> sql("select grouping__id from t3").show()
org.apache.spark.sql.AnalysisException: grouping_id() can only be used with GroupingSets/Cube/Rollup;;

// w/ this PR
scala> sql("select grouping__id from t3").show()
+------------+
|grouping__id|
+------------+
|           3|
+------------+
```

Reply:
This reminds me of `Analyzer.resolveLiteralFunction`. For these identifier-like functions, we should try to resolve them as columns first.

Personally, I find the identifier-like functions a very tricky syntax. `CURRENT_DATE` etc. is from the SQL standard so I'm fine with it, plus we treat `CURRENT_DATE` as a reserved keyword under ANSI mode. `grouping__id` is not standard and I'm wondering if we should just forbid it (with a legacy config). Users must use `grouping__id()`.

Reply:
IIUC `grouping__id` exists for hive interoperability? Anyway, the idea looks reasonable. I'll make a PR to follow it.

Reply:
https://github.com/apache/spark/pull/31751